### PR TITLE
[Core] Standalone PR to add the class OBB

### PR DESCRIPTION
**Pull request by loumalouomega**
This PR is part of #4263 and adds the OBB class

### CODE REVIEW

**Comment by maceligueta** [comment]
Shouldn't we use more explicative names? It could be `oriented_bounding_box.cpp` (in case I guessed correctly).

**Comment by loumalouomega** [comment]
DOne

File: kratos/includes/oriented_bounding_box.h [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+ * @details The geometrical definition of the OBB can be done as the
+ *                      ` *
+ *            directions / `
+ *                  * \ / half diagonals
+ *                  `  *C  *
+ *                  ` / \ `
+ *                  `/  `
+ *                  * `
+ * For more details https://www.geometrictools.com/Documentation/DynamicCollisionDetection.pdf
+ * @tparam
```
Comment (loumalouomega) [inline kratos/includes/oriented_bounding_box.h]:
```suggestion
 * @tparam TDim The dimension of the space
```

**Comment by adityaghantasala** [comment]
if possible also changing the name of the class from OBB will also make it more clear !

**Comment by loumalouomega** [comment]
OK...

**Comment by loumalouomega** [comment]
Done

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
Some comments on the first view:
- The constructor arguments (and internal data) can be reduced by getting 4 points. Is there any special reason to have it with center,vectors and half lengths?
- There are typos (like OutpuType ) and styling problems
- Are you sure if your IsInside algorithm is optimal? What is the reference?

**Comment by loumalouomega** [comment]
> Some comments on the first view:
> 
>     * The constructor arguments (and internal data) can be reduced by getting 4 points. Is there any special reason to have it with center,vectors and half lengths?
> 
I use as reference several implementations found on Internet:
https://github.com/juj/MathGeoLib/blob/master/src/Geometry/OBB.h
https://github.com/juj/MathGeoLib/blob/master/src/Geometry/OBB.cpp
>     * There are typos (like OutpuType ) and styling problems
I will fix it
> 
>     * Are you sure if your IsInside algorithm is optimal? What is the reference?
I could not found any specific algorithm for intersection, but for "IsInside". Checking the IsInside for each point of the opposite object is the faster way (check previous OBB.h/cpp files), but doesn't cover 100% of cases, then you need to check the faces are intersected.

**Comment by loumalouomega** [comment]
@pooyan-dadvand I don't know how can I generalize the definition of the OOB using only points for the 2D and 3D OBB

**Comment by pooyan-dadvand** [comment]
One as base point and 3 defining the major axis and length for each local axis. 

And please check the intersection algorithms in literature. There are many of them and I don't have experience in OOBB for recommending one. You may find some references here:

http://www.realtimerendering.com/intersections.html

**Comment by loumalouomega** [comment]
> One as base point and 3 defining the major axis and length for each local axis.
> 
> And please check the intersection algorithms in literature. There are many of them and I don't have experience in OOBB for recommending one. You may find some references here:
> 
> http://www.realtimerendering.com/intersections.html

OK, I will

**Comment by loumalouomega** [comment]
Looking a little bit more I found the separating axis theorem:
https://gamedev.stackexchange.com/questions/112883/simple-3d-obb-collision-directx9-c
https://gamedevelopment.tutsplus.com/tutorials/collision-detection-using-the-separating-axis-theorem--gamedev-169

**Comment by loumalouomega** [comment]
@pooyan-dadvand I implemented the separating axis theorem and I consider it by default

**Comment by pooyan-dadvand** [comment]
Why are you keeping the direct one by an enum? Is there any case that would be better than the separation of axis?

**Comment by loumalouomega** [comment]
> Why are you keeping the direct one by an enum? Is there any case that would be better than the separation of axis?

It could take less time in very specific cases. Separation axis requires to do all the tests always. Direct approach can be faster if one node is inside the other geometry (require less operations in this specific case)

**Comment by pooyan-dadvand** [comment]
> It could take less time in very specific cases. Separation axis requires to do all the tests always. Direct approach can be faster if one node is inside the other geometry (require less operations in this specific case)

So we cannot anticipate if the direct one is faster. I would definitely keep the separation of axis only.

**Comment by loumalouomega** [comment]
> > It could take less time in very specific cases. Separation axis requires to do all the tests always. Direct approach can be faster if one node is inside the other geometry (require less operations in this specific case)
> 
> So we cannot anticipate if the direct one is faster. I would definitely keep the separation of axis only.

The only problem I have with the axis projection is that it makes always all the checks, instead of incrementally

**Comment by pooyan-dadvand** [comment]
I checked your implementation and it is incremental. Considering that a chain of or operators are always evaluated to true when the first true is found and the rest of the statements won't be evaluated.

File: kratos/includes/oriented_bounding_box.h [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+        ) const;
+
+    /**
+     * @brief Fast compute of cross product without considering checks. Performs the cross product of the two input vectors a,b
+     * @details a,b are assumed to be of size 3
+     * @param a First input vector
+     * @param b Second input vector
+     * @return c The resulting vector
+     */
+    static array_1d<double, 3> FastCrossProduct(const array_1d<double, 3>& a, const array_1d<double, 3>& b)
```
Comment (pooyan-dadvand) [inline kratos/includes/oriented_bounding_box.h]: Why we need this method? Shouldn't be in math utils?

Comment (loumalouomega) [reply kratos/includes/oriented_bounding_box.h]: Could be

Comment (loumalouomega) [reply kratos/includes/oriented_bounding_box.h]: The only difference is that t skips the checks

Comment (pooyan-dadvand) [reply kratos/includes/oriented_bounding_box.h]: The tests are in debug only. So please use the math utils one

Comment (loumalouomega) [reply kratos/includes/oriented_bounding_box.h]: OK

**Comment by loumalouomega** [comment]
Done @pooyan-dadvand

**Comment by loumalouomega** [comment]
Ping

**Review by pooyan-dadvand — APPROVED** [review]
Thanks for considering the comments.

**Comment by loumalouomega** [comment]
Thanks

**Pull request merged**